src/python/torchdistx/gossip_grad.py

```diff
@@ -0,0 +1,338 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import random
+from enum import Enum, auto
+from itertools import cycle
+
+import torch
+import torch.distributed as dist
+import torch.distributed.distributed_c10d as c10d
+from torch._C._distributed_c10d import ProcessGroup
+from torch.distributed.algorithms._comm_hooks import default
+
+
+class Topology(Enum):
+    r"""
+    Specifies which topology will be used as a base for gradient communication.
+    For more information, please refer to the original
+    `paper <https://arxiv.org/abs/1803.05880>`_
+
+    CUBE: A hypercube topology - a hierarchical virtual organization of compute nodes.
+        For this topology gossiping is happening with a neighboring vertex.
+              *----*
+             /|   /|
+            *----* |
+            | * -|-*
+            |/   |/
+            *----*
+
+    DISSEMINATION: A dissemination topology has similar property
+        as hypercube virtual topology.
+        For this topology gossiping is happening with the neighboring node,
+        then every 2nd node, every 4th, etc.
+
+            .  *  .
+          *          *
+        .              .
+        *              *
+        .              .
+          *          *
+            .  *  .
+
+    .. note::
+        Current implementation does not support uneven number of nodes for a CUBE
+        topology.
+
+    """
+    CUBE = auto()
+    DISSEMINATION = auto()
+
+
+class GossipGraDState(default.DefaultState):
+    r"""
+    Stores state needed to perform GossipGraD algorithm within a communication hook.
+
+    Args:
+        topology (Topology): A virtual topology to be used for gradient communication.
+            (default: DISSEMINATION)
+        local_process_group (ProcessGroup): Stores local subgroup,
+            where intra-node communication will happen,
+            by default a subgroup is initialized to workers, belonging to the same node.
+            Should be provided together with `num_nodes`.
+            (default: None)
+        num_nodes (int): Number of nodes in a compute environment.
+            Should be provided together with `local_process_group`.
+            By default is initialized to the number of generated local subgroups.
+            (default: None)
+        master_process_group (ProcessGroup): Stores main workers,
+            which are involved in inter-node communication. By default, will be
+            composed from the workers with ranks 0 in the local process group.
+            (default: None)
+
+    """
+
+    def __init__(
+        self,
+        topology=None,
+        local_process_group=None,
+        num_nodes=None,
+        master_process_group=None,
+        proc_per_node=None,
```

Comment:
they are useful, if a user wants to create his own environment, e.g. simulate a presence of multiple nodes on a 1 node environment, then he can create his own `local_process_group` and `master_process_group` and provide the rest of args. Otherwise, if nothing is provided, they will be set to default